### PR TITLE
fix for the server crash when the maxclients increased via config set

### PR DESCRIPTION
**Pull request by surenkajan**
multi threaded KeyDB server  crashed when we increased the maxclients via CONFIG SET command.

**Ways to reproduce this error:**

set the maxclients to 20000 in keydb.conf

start the keydb-server (server-threads = 4)

open the keydb-cli and issue "CONFIG SET maxclients 30000"

keydb-server will crash immediately with core dumb file with below error message:
AE_ASSERT FAILURE ae.cpp: 392

**Solution**

Use the _aePostFunction_ to correctly post the request to corresponding worker thread.

This PR will solve this crash issue.

### CODE REVIEW

File: src/config.cpp [hunk 1]
```diff
@@ -2278,10 +2278,14 @@ static int updateMaxclients(long long val, long long prev, const char **err) {
             if ((unsigned int) aeGetSetSize(g_pserver->rgthreadvar[iel].el) <
                 g_pserver->maxclients + CONFIG_FDSET_INCR)
             {
-                if (aeResizeSetSize(g_pserver->rgthreadvar[iel].el,
-                    g_pserver->maxclients + CONFIG_FDSET_INCR) == AE_ERR)
-                {
-                    *err = "The event loop API used by Redis is not able to handle the specified number of clients";
+                int res = aePostFunction(g_pserver->rgthreadvar[iel].el, [iel] {
+                    aeResizeSetSize(g_pserver->rgthreadvar[iel].el, g_pserver->maxclients + CONFIG_FDSET_INCR);
```
Comment (JohnSully) [inline src/config.cpp:2282]: At a minimum we need to do something useful with the aeResizeSetSize() return code for teh current thread.  Checking the return of aePostFunction is not sufficient to know the update will actually work.

Comment (surenkajan) [reply src/config.cpp]: Thanks for your feedback John.

aePostFunction (with default values) will always return "**AE_OK**" for the current thread  after it successfully call the "aeResizeSetSize" function. It seems there is no way to get the return code for aeResizeSetSize call. 

I tried to call aePostFunction with "synchronous = true" , but, client(cli) waits forever for the reply.  Is there  a way to get immediate execution on each event-loop? 

However, Each threads can check the return code of "aeResizeSetSize" call. Based on that, I implemented the error handling. Please let me know your thoughts on this changes.

Comment (surenkajan) [reply src/config.cpp]: @JohnSully aeResizeSetSize() return code for the current thread is properly handled while serverLog is updated for all other threads.

Please let me know your thoughts on this changes

File: src/config.cpp [hunk 2]
```diff
@@ -2273,17 +2273,43 @@ static int updateMaxclients(long long val, long long prev, const char **err) {
             }
             return 0;
         }
+	/* Change the SetSize for the current thread first. 
+	 * If any error, return the error message to the client, otherwise, continue to do the same for other threads */
+        if ((unsigned int) aeGetSetSize(aeGetCurrentEventLoop()) <
```
Comment (JohnSully) [inline src/config.cpp:2278]: Just a few formatting issues and we'll be good to go:

1) Your mixing tabs and spaces resulting in inconsistent indentation
2) Use this style for new code:
if (x > y) {
   ...
}
3) Don't try to keep lines 80 chars wide, modern monitors are much wider than tall.  Use judgement when to spill to a new line, but in general the old 80 char standard is far too narrow.

Comment (surenkajan) [reply src/config.cpp:2278]: Thanks John for the feedback. 
I corrected the format issues as you suggested.

File: src/config.cpp [hunk 3]
```diff
@@ -2273,17 +2273,43 @@ static int updateMaxclients(long long val, long long prev, const char **err) {
...
+            {
+                *err = "The event loop API used by Redis is not able to handle the specified number of clients";
+                return 0;
+            }
+	    serverLog(LL_DEBUG,"Successfully changed the setsize for current thread %d", ielFromEventLoop(aeGetCurrentEventLoop()));
+        }
+
         for (int iel = 0; iel < cserver.cthreads; ++iel)
         {
+	    if (g_pserver->rgthreadvar[iel].el == aeGetCurrentEventLoop())
```
Comment (JohnSully) [inline src/config.cpp:2292]: Still a tab here

File: src/config.cpp [hunk 4]
```diff
@@ -2273,17 +2273,43 @@ static int updateMaxclients(long long val, long long prev, const char **err) {
...
+                    }
+                });
+
+                if (res != AE_OK){
+                    static char msg[128];
+                    sprintf(msg, "Failed to post the request to change setsize for Thread %d", iel);
+                    *err = msg;
                     return 0;
                 }
+		serverLog(LL_DEBUG,"Successfully post the request to change the setsize for thread %d", iel);
```
Comment (JohnSully) [inline src/config.cpp:2312]: Still tabs

File: src/config.cpp [hunk 5]
```diff
@@ -2273,17 +2273,43 @@ static int updateMaxclients(long long val, long long prev, const char **err) {
...
+                *err = "The event loop API used by Redis is not able to handle the specified number of clients";
+                return 0;
+            }
+	    serverLog(LL_DEBUG,"Successfully changed the setsize for current thread %d", ielFromEventLoop(aeGetCurrentEventLoop()));
+        }
+
         for (int iel = 0; iel < cserver.cthreads; ++iel)
         {
+	    if (g_pserver->rgthreadvar[iel].el == aeGetCurrentEventLoop())
+            {
```
Comment (JohnSully) [inline src/config.cpp:2293]: We're mixing brace styles in this change still.

Lines you touch should be converted to:
if (x > y) {
   ...
}

Note I'm only putting one comment in but there are multiple lines with this issue.

**Review by JohnSully — CHANGES_REQUESTED** [review]

File: src/config.cpp (hunk 3 above)
Comment (surenkajan) [reply src/config.cpp]: Fixed.

File: src/config.cpp (hunk 4 above)
Comment (surenkajan) [reply src/config.cpp]: Fixed.

File: src/config.cpp (hunk 5 above)
Comment (surenkajan) [reply src/config.cpp]: Fixed.

**Comment by surenkajan** [comment]
@JohnSully Thanks for your feedback and I addressed all your comments in the last commit. 

Please have a look and let me know if you have any concerns on merging

**Comment by JohnSully** [comment]
Thanks for your help with this! I really appreciate it.

**Comment by surenkajan** [comment]
Thanks John.

**Pull request merged**